src/Stratis.Bitcoin.Features.SmartContracts/Consensus/Rules/SmartContractPosCoinviewRule.cs

```diff
@@ -0,0 +1,367 @@
+
+            this.blockTxsProcessed = new List<Transaction>();
+            this.refundCounter = 1;
+
+            this.CheckAndComputeStake(context);
+
+            await base.RunAsync(context).ConfigureAwait(false);
+            var posRuleContext = context as PosRuleContext;
+            await this.stakeChain.SetAsync(context.ValidationContext.ChainedHeader, posRuleContext.BlockStake).ConfigureAwait(false);
+
+            this.Logger.LogTrace("(-)");
+        }
+
+        /// <inheritdoc/>
+        protected override bool IsProtocolTransaction(Transaction transaction)
+        {
+            return transaction.IsCoinBase || transaction.IsCoinStake;
+        }
+
+        /// <inheritdoc />
+        public override void CheckBlockReward(RuleContext context, Money fees, int height, NBitcoin.Block block)
+        {
+            this.Logger.LogTrace("({0}:{1},{2}:'{3}')", nameof(fees), fees, nameof(height), height);
+
+            if (BlockStake.IsProofOfStake(block))
+            {
+                var posRuleContext = context as PosRuleContext;
+                Money stakeReward = block.Transactions[1].TotalOut - posRuleContext.TotalCoinStakeValueIn;
+                Money calcStakeReward = fees + this.GetProofOfStakeReward(height);
+
+                this.Logger.LogTrace("Block stake reward is {0}, calculated reward is {1}.", stakeReward, calcStakeReward);
+                if (stakeReward > calcStakeReward)
+                {
+                    this.Logger.LogTrace("(-)[BAD_COINSTAKE_AMOUNT]");
+                    ConsensusErrors.BadCoinstakeAmount.Throw();
+                }
+            }
+            else
+            {
+                Money blockReward = fees + this.GetProofOfWorkReward(height);
+                this.Logger.LogTrace("Block reward is {0}, calculated reward is {1}.", block.Transactions[0].TotalOut, blockReward);
+                if (block.Transactions[0].TotalOut > blockReward)
+                {
+                    this.Logger.LogTrace("(-)[BAD_COINBASE_AMOUNT]");
+                    ConsensusErrors.BadCoinbaseAmount.Throw();
+                }
+            }
+
+            this.Logger.LogTrace("(-)");
+        }
+
+        /// <inheritdoc />
+        public override void UpdateCoinView(RuleContext context, Transaction transaction)
+        {
+            this.Logger.LogTrace("()");
+
+            if (this.generatedTransaction != null)
+            {
+                ValidateGeneratedTransaction(transaction);
+                base.UpdateUTXOSet(context, transaction);
+                return;
+            }
+
+            // If we are here, was definitely submitted by someone
+            ValidateSubmittedTransaction(transaction);
+
+            TxOut smartContractTxOut = transaction.Outputs.FirstOrDefault(txOut => txOut.ScriptPubKey.IsSmartContractExec());
+            if (smartContractTxOut == null)
+            {
+                // Someone submitted a standard transaction - no smart contract opcodes.
```

Comment:
feels like methods appropriate for this standard vs smart contract transaction, and possibly even a strategy injected instead of the conditional checks.